File: LeetCode/1802.Maximum_Value_at_a_Given_Index_in_a_Bounded_Array/Mario/1802.Maximum_Value_at_a_Given_Index_in_a_Bounded_Array.py
#!/usr/bin/env python3


class Solution:
    def maxValue(self, n: int, index: int, maxSum: int) -> int:

        def sideSum(m, l):
            if m - 1 >= l:
                return (m - 1 + m - l) * l // 2
            else:
                return (m - 1 + 1) * (m - 1) // 2 + l - (m - 1)

        lo, hi = 1, maxSum + 1
        res = 1
        while lo < hi:
            mid = (lo + hi) // 2

            total = sideSum(mid, index) + sideSum(mid, n - index - 1) + mid
            if total <= maxSum:
                res = mid
                lo = mid + 1
            else:
                hi = mid

        return res

File: LeetCode/1802.Maximum_Value_at_a_Given_Index_in_a_Bounded_Array/1802.Maximum_Value_at_a_Given_Index_in_a_Bounded_Array.swift
class Solution {
    func sum(_ k: Int, _ i: Int, _ n: Int) -> Int {
        let lcnt = i, rcnt = n - 1 - i

        func sideSum(_ cnt: Int) -> Int {
            if cnt < k - 1 {
                return (k - 1 + k - cnt) * cnt / 2
            } else {
                return k * (k - 1) / 2 + cnt - (k - 1)
            }
        }

        return sideSum(lcnt) + sideSum(rcnt) + k
    }

    func maxValue(_ n: Int, _ index: Int, _ maxSum: Int) -> Int {
        // binary search
        var l = 1, r = maxSum + 1
        var res = 0
        while l < r {
            let mid = (l + r) >> 1
            if sum(mid, index, n) <= maxSum {
                l = mid + 1
                res = mid
            } else {
                r = mid
            }
        }

        return res
    }
}
